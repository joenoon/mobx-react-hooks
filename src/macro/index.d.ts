// Provides typings for the macro version of useObserver

// a little duplication to avoid tsc building things outside this directory
type ForceUpdateHook = () => () => void
interface IUseObserverOptions {
    useForceUpdate?: ForceUpdateHook
}

// note: different signature than regular useObserver
export declare function useObserver<T>(
    baseComponentName?: string,
    options?: IUseObserverOptions
): void
