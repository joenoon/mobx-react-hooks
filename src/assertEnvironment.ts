import { spy } from "mobx"
import { useState } from "react"

if (!useState) {
    throw new Error("@joenoon/mobx-react-hooks requires React with Hooks support")
}
if (!spy) {
    throw new Error("@joenoon/mobx-react-hooks requires mobx at least version 4 to be available")
}
