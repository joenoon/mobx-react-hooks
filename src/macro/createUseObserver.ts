// @ts-ignore
import { createMacro } from "babel-plugin-macros"
import path from "path"

interface ICreateUseObserverConfig {
    /**
     * provide an absolute path if this file is in your application, OR
     * provide a library import name.
     * this is transformed to:
     *   import { importSpecifier } from 'relative path to importSource, or as-is for library name';
     */
    importSource: string
    /**
     * provide the exported name from your importSource. i.e.: useMyObserverFunc
     * this is transformed to:
     *   import { importSpecifier } from 'relative path to importSource, or as-is for library name';
     */
    importSpecifier: string
}

export function createUseObserver(config: ICreateUseObserverConfig) {
    return createMacro(function useObserverMacro({
        references,
        state: {
            file: {
                opts: { filename }
            }
        },
        babel: { types: t }
    }: any) {
        // transform the importSource to a relative path if its not a library
        let { importSource } = config
        if (path.isAbsolute(importSource)) {
            importSource = `./${path.relative(path.dirname(filename), importSource)}`
        }

        const { useObserver = [] } = references

        useObserver.forEach((referencePath: any) => {
            // program is the top scope of the file
            const program = referencePath.scope.getProgramParent().path
            const line = referencePath.getStatementParent()
            const lineArgs = referencePath.container.arguments
            // get all lines in the scope that follow.  these get moved automatically
            const linesAfter = line.container.splice(
                line.key + 1,
                line.container.length - line.key - 1
            )

            // return a new function containing linesAfter.
            // the name `useObserverRenderHook` is not important.
            line.insertAfter(
                t.returnStatement(
                    t.callExpression(t.identifier(config.importSpecifier), [
                        t.functionExpression(
                            t.identifier("useObserverRenderHook"),
                            [],
                            t.blockStatement(linesAfter),
                            false,
                            false
                        ),
                        ...lineArgs
                    ])
                )
            )
            // remove the original line
            line.remove()

            // check to see if we've already added the real import line to this file
            const foundImport =
                program.get("body").filter((x: any) => {
                    if (t.isImportDeclaration(x) && x.node.source.value === importSource) {
                        return (
                            x.node.specifiers.filter(
                                (s: any) => s.local.name === config.importSpecifier
                            ).length > 0
                        )
                    }
                    return false
                }).length > 0

            // if not, add it, i.e.:
            // import { importSpecifier } from importSource
            if (!foundImport) {
                program.unshiftContainer(
                    "body",
                    t.importDeclaration(
                        [
                            t.importSpecifier(
                                t.identifier(config.importSpecifier),
                                t.identifier(config.importSpecifier)
                            )
                        ],
                        t.stringLiteral(importSource)
                    )
                )
            }
        })
    })
}
