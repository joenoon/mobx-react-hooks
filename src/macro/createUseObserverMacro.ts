// @ts-ignore
import { createMacro } from "babel-plugin-macros"
import path from "path"

interface ICreateUseObserverMacroConfig {
    /**
     * provide an absolute path if this file is in your application, OR
     * provide a library import name.
     * this is transformed to:
     *   import { importSpecifier } from 'relative path to importSource, or as-is for library name';
     */
    importSource: string
    /**
     * provide the exported name from your importSource. i.e.: useMyObserverFunc
     * this is transformed to:
     *   import { importSpecifier } from 'relative path to importSource, or as-is for library name';
     */
    importSpecifier: string
}

/**
 *
 */

/**
 * Set up as follows:
 *
 * @example
 * ```javascript
 * // myapp/src/lib/customUseObserver.macro.js:
 *
 * const { createUseObserverMacro } = require("mobx-react-lite/macro/createUseObserverMacro")
 * const path = require("path")
 * module.exports = createUseObserverMacro({
 *     importSource: path.resolve(__dirname, "./customUseObserver"),
 *     importSpecifier: "useCustomObserver"
 * })
 * ```
 *
 * @example
 * ```typescript
 * // myapp/src/lib/customUseObserver.ts:
 *
 * import { useComputed, useObserver, useObservable } from "mobx-react-lite"
 *
 * const HOOKS = {
 *     useComputed,
 *     useObservable,
 *     observerHooks: {
 *         useComputed,
 *         useObservable
 *     }
 * }
 *
 * export type UseCustomObserverHooks = typeof HOOKS
 *
 * export function useCustomObserver<T>(
 *     func: (hooks: UseCustomObserverHooks) => T,
 *     baseComponentName?: string
 * ): T {
 *     return useObserver(() => func(HOOKS), baseComponentName)
 * }
 * ```
 *
 * @example
 * ```typescript
 * // myapp/src/lib/customUseObserver.ts.d:
 *
 * import { useComputed, useObservable } from "mobx-react-lite"
 * declare module "useCustomObserver.macro"
 *
 * interface UseCustomObserverHooks {
 *     useComputed: typeof useComputed
 *     useObservable: typeof useObservable
 *     observerHooks: UseCustomObserverHooks
 * }
 *
 * export declare function useObserver<T>(baseComponentName?: string): UseCustomObserverHooks
 * ```
 *
 * @example
 * ```typescript
 * //Then, to use it in your app:
 *
 * import { useObserver } from "../path/to/useCustomObserver.macro"
 *
 * export const MyComponent = props => {
 *     const { useObservable, useComputed } = useObserver()
 *     const foo = useObservable({ foo: 1 })
 *     const bar = useComputed(() => foo.foo + 1)
 *     return <div>...</div>
 * }
 * ```
 */
export function createUseObserverMacro(config: ICreateUseObserverMacroConfig) {
    return createMacro(function useObserverMacro({
        references,
        state: {
            file: {
                opts: { filename }
            }
        },
        babel: { types: t }
    }: any) {
        // transform the importSource to a relative path if its not a library
        let { importSource } = config
        if (path.isAbsolute(importSource)) {
            importSource = `./${path.relative(path.dirname(filename), importSource)}`
        }

        const { useObserver = [] } = references

        useObserver.forEach((referencePath: any) => {
            // program is the top scope of the file
            const program = referencePath.scope.getProgramParent().path
            const line = referencePath.getStatementParent()
            // lineVars is the `{a, b}` in: const {a, b} = useObserver();
            const lineVars = line.node.declarations ? line.node.declarations[0].id : null
            // get all lines in the scope that follow.  these get moved automatically
            const linesAfter = line.container.splice(
                line.key + 1,
                line.container.length - line.key - 1
            )

            // return a new function taking lineVars as args, containing linesAfter.
            // the name `useObserverRenderHook` is not important.
            line.insertAfter(
                t.returnStatement(
                    t.callExpression(t.identifier(config.importSpecifier), [
                        t.functionExpression(
                            t.identifier("useObserverRenderHook"),
                            lineVars ? [lineVars] : [],
                            t.blockStatement(linesAfter),
                            false,
                            false
                        )
                    ])
                )
            )
            // remove the original line
            line.remove()

            // check to see if we've already added the real import line to this file
            const foundImport =
                program.get("body").filter((x: any) => {
                    if (t.isImportDeclaration(x) && x.node.source.value === importSource) {
                        return !!x.node.specifiers.filter(
                            (s: any) => s.local.name === config.importSpecifier
                        )
                    }
                    return false
                }).length > 0

            // if not, add it, i.e.:
            // import { importSpecifier } from importSource
            if (!foundImport) {
                program.unshiftContainer(
                    "body",
                    t.importDeclaration(
                        [
                            t.importSpecifier(
                                t.identifier(config.importSpecifier),
                                t.identifier(config.importSpecifier)
                            )
                        ],
                        t.stringLiteral(importSource)
                    )
                )
            }
        })
    })
}
