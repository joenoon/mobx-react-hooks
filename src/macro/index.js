const { createUseObserver } = require("./createUseObserver")
const path = require("path")
module.exports = createUseObserver({
    importSource: "mobx-react-lite",
    importSpecifier: "useObserver"
})
