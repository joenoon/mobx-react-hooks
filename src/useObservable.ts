import { observable } from "mobx"
import { useState } from "react"

type SupportedValues = object | Map<any, any> | any[]

export function useObservable<T extends SupportedValues>(initialValue: T | (() => T)): T {
    const fn = () =>
        observable(typeof initialValue === "function" ? (initialValue as any)() : initialValue)
    return useState(fn)[0]
}
