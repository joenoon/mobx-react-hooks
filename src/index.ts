import "./assertEnvironment"

export { isUsingStaticRendering, useStaticRendering } from "./staticRendering"
export { useObserverInternal, ForceUpdateHook, IUseObserverOptions } from "./useObserverInternal"
export { useForceUpdate } from "./utils"
