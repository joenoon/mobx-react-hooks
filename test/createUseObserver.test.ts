// @ts-ignore
import plugin from "babel-plugin-macros"
// @ts-ignore
import pluginTester from "babel-plugin-tester"

pluginTester({
    plugin,
    snapshot: true,
    babelOptions: { filename: __filename },
    tests: [
        `
      import {useObserver} from '../src/macro'

      export const MyComponent = (props) => {
        useObserver();
        const foo = 1;
        const bar = 2;
        return null;
      };

      export const MyComponent2 = (props) => {
        // a comment
        useObserver();
        const foo = 1;
        return null;
      };

      export const MyComponent3 = (props) => {
        useObserver();
        return null;
      };

      export const MyComponent4 = (props) => {
        useObserver('baseComponentName');
        return null;
      };

      export const MyComponent5 = (props) => {
        useObserver('baseComponentName', { useForceUpdate: true });
        return null;
      };

      export const MyComponent6 = (props) => {

        useObserver();
        const foo = 1;
        return null;
      };
      `,
        {
            error: true,
            code: `
      import {useObserver} from '../src/macro'

      export const MyComponent7 = (props) => {
        dontDoThis();
        useObserver();
        const foo = 1;
        return null;
      };
      `
        }
    ]
})
