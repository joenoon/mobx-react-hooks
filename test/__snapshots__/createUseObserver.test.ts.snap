// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros 1. macros: 1. macros 1`] = `
"
import {useObserver} from '../src/macro'

export const MyComponent = (props) => {
  useObserver();
  const foo = 1;
  const bar = 2;
  return null;
};

export const MyComponent2 = (props) => {
  // a comment
  useObserver();
  const foo = 1;
  return null;
};

export const MyComponent3 = (props) => {
  useObserver();
  return null;
};

export const MyComponent4 = (props) => {
  useObserver('baseComponentName');
  return null;
};

export const MyComponent5 = (props) => {
  useObserver('baseComponentName', { useForceUpdate: true });
  return null;
};

export const MyComponent6 = (props) => {

  useObserver();
  const foo = 1;
  return null;
};

      ↓ ↓ ↓ ↓ ↓ ↓

import { useObserverInternal } from \\"@joenoon/mobx-react-hooks\\";
export const MyComponent = props => {
  return useObserverInternal(function useObserverRenderHook() {
    const foo = 1;
    const bar = 2;
    return null;
  });
};
export const MyComponent2 = props => {
  // a comment
  return useObserverInternal(function useObserverRenderHook() {
    const foo = 1;
    return null;
  });
};
export const MyComponent3 = props => {
  return useObserverInternal(function useObserverRenderHook() {
    return null;
  });
};
export const MyComponent4 = props => {
  return useObserverInternal(function useObserverRenderHook() {
    return null;
  }, 'baseComponentName');
};
export const MyComponent5 = props => {
  return useObserverInternal(function useObserverRenderHook() {
    return null;
  }, 'baseComponentName', {
    useForceUpdate: true
  });
};
export const MyComponent6 = props => {
  return useObserverInternal(function useObserverRenderHook() {
    const foo = 1;
    return null;
  });
};
"
`;

exports[`macros 2. macros: 2. macros 1`] = `
"
import {useObserver} from '../src/macro'

export const MyComponent7 = (props) => {
  dontDoThis();
  useObserver();
  const foo = 1;
  return null;
};

      ↓ ↓ ↓ ↓ ↓ ↓

MacroError: createUseObserver.test.ts:5: useObserver() provided by the macro must be the first line of the FunctionComponent.
"
`;
